docs/src/counterintuitive.md

```diff
@@ -1,42 +1,192 @@
+
+```jldoctest counter1
+julia> soa
+4-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Foo:
+ Foo{Int64}(5, 2)
+ Foo{Int64}(6, 7)
+ Foo{Int64}(1, 8)
+ Foo{Int64}(1, 2)
+
+julia> a
+4-element Vector{Int64}:
+ 5
+ 6
+ 1
+ 1
+
+julia> b
+4-element Vector{Int64}:
+ 2
+ 7
+ 8
+ 2
+```
+
+This is because `soa` is a "view" of `a` and `b` (it has no independent storage of its own).
+
+However, you may be surprised by the following:
+
+```jldoctest counter1
+julia> soa[4].b = 9
+9
+
+julia> soa
+4-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Foo:
+ Foo{Int64}(5, 2)
+ Foo{Int64}(6, 7)
+ Foo{Int64}(1, 8)
  Foo{Int64}(1, 2)
+
+julia> b
+4-element Vector{Int64}:
+ 2
+ 7
+ 8
+ 2
+```
+
+This assignment had no persistent effect on `soa` or `b`. This occurs because `soa[4]` is generated
+on-the-fly; since it returns a `Foo`, which is mutable, you can change its fields. However, the modified `Foo` object
+is not stored anywhere.
+
+To store a modification, one would instead need
+
+```jldoctest counter1
+julia> x = soa[4]; x.b = 10; soa[4] = x     # store the modified `x` in `soa`
+Foo{Int64}(1, 10)
+
+julia> soa
+4-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Foo:
+ Foo{Int64}(5, 2)
+ Foo{Int64}(6, 7)
+ Foo{Int64}(1, 8)
+ Foo{Int64}(1, 10)
+
+julia> b
+4-element Vector{Int64}:
+  2
+  7
+  8
+ 10
 ```
-The assignment `x[1].a = 10` first calls `getindex(x,1)`, then sets property `a` of the accessed element. However, since StructArrays constructs `Foo(x.a[1],x.b[1])` on the fly when when accessing `x[1]`, setting `x[1].a = 10` modifies the materialized struct rather than the StructArray `x`. 
 
-Note that one can modify a field of a StructArray entry via `x.a[1] = 10` (the order of `.` syntax and indexing syntax matters). As an added benefit, this does not require that the struct `Foo` is mutable, as it modifies the underlying component array `x.a` directly.
+!!! note
+    This behavior only arises for *mutable* `struct`s. If `Foo` were immutable, re-assigning the `b` field would be an error, and there would be no opportunity for confusion. Moreover, the performance of immutable struct creation is generally much better than for mutable structs. Thus, it is recommended to use immutable structs with StructArray whenever possible.
+
+### The "copy" case (AOS->SOA)
+
+Above, we created a StructArray from arrays `a` and `b`, which creates a "view." The same is not true if you create a StructArray from an array-of-structs:
+
+```jldoctest counter1
+julia> aos = [Foo(1,2) for i = 1:4]
+4-element Vector{Foo{Int64}}:
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+
+julia> soa = StructArray(aos)
+4-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Foo{Int64}:
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+ Foo{Int64}(1, 2)
+
+julia> soa.a[1] = 5
+5
+
+julia> soa[2] = Foo(6, 7)
+Foo{Int64}(6, 7)
+
+julia> b[3] = 8
```

Comment:
```suggestion
```
Probably shouldn't mention `b` since it's not involved in the construction of `aos`.